pkg/kube/object.go

```diff
@@ -253,8 +256,66 @@ var ErrReplicaSetNotFound = errors.New("replicaset not found")
 var ErrNoRunningPods = errors.New("no active pods for controller")
 var ErrUnSupportedKind = errors.New("unsupported workload kind")
 
+//CompatibleMgr provide k8s compatible objects (group/api/kind) capabilities
+type CompatibleMgr interface {
+	// GetSupportedObjectByKind get specific k8s compatible object (group/api/kind) by kind
+	GetSupportedObjectByKind(kind Kind) client.Object
+}
+
+type CompatibleObjectMapper struct {
+	kindObjectMap map[string]client.Object
+}
+
 type ObjectResolver struct {
 	client.Client
+	CompatibleMgr CompatibleMgr
+}
+
+func NewObjectResolver(c client.Client, cm CompatibleMgr) ObjectResolver {
+	return ObjectResolver{c, cm}
+}
+
+//InitCompatibleMgr initialized a CompatibleObjectMapper who store a map the of supported kinds with it compatible Objects (group/api/kind)
+// it dynamically fetches the compatible k8s objects (group/api/kind) by resource from the cluster and store it in kind vs k8s object mapping
+// It will enable the operator to support old and new API resources based on cluster version support
+func InitCompatibleMgr(restMapper meta.RESTMapper) (CompatibleMgr, error) {
+	for _, resource := range getCompatibleResources() {
+		gvk, err := restMapper.KindFor(schema.GroupVersionResource{Resource: resource})
+		if err != nil {
+			return nil, err
+		}
+		sok, err := supportedObjectsByK8sKind(gvk.String())
+		if err != nil {
+			return nil, err
+		}
+		return &CompatibleObjectMapper{kindObjectMap: sok}, nil
+	}
+	return nil, fmt.Errorf("compatible Mgr could not be initilized")
+}
+
+// return a map of supported object api per k8s version
+func supportedObjectsByK8sKind(api string) (map[string]client.Object, error) {
+	kindObjectMap := make(map[string]client.Object)
+	var resource client.Object
+	switch api {
+	case apiBatchV1beta1CronJob:
+		resource = &batchv1beta1.CronJob{}
+	case apiBatchV1CronJob:
+		resource = &batchv1.CronJob{}
+	default:
+		return nil, fmt.Errorf("api %s is not suooprted compatibale resource", api)
+	}
+	kindObjectMap[string(KindCronJob)] = resource
```

Comment:
This breaks the idea here for me, because we have `InitCompatibleMgr` doing a range on compatible types, and then calling `supportedObjectsByK8sKind` to return the map per `gvk`, but inside `supportedObjectsByK8sKind` we hardcode `resource` to a `KindCronJob`

Reply:
this the most flexible you can get as you need to provide [object type](https://github.com/aquasecurity/trivy-operator/blob/main/pkg/configauditreport/controller/controller.go#L98) , but this whey we do not need to be depend on version , we know where is the computability problem so it can be managed